src/sycltest/plugin-Test1/gpuAlgo1.cc

```diff
@@ -0,0 +1,175 @@
+  // query the device for the maximum "block size" or workgroup size
+  const unsigned int workgroupSize = stream.get_device().get_info<sycl::info::device::max_work_group_size>();
+
+  // build a rectangular "block" or workgroup
+  const unsigned int workgroupRectB = sqrt2(workgroupSize);
+  const unsigned int workgroupRectA = workgroupSize / workgroupRectB;
+
+  auto h_a = cms::sycltools::make_host_unique<float[]>(NUM_VALUES, stream);
+  auto h_b = cms::sycltools::make_host_unique<float[]>(NUM_VALUES, stream);
+
+  for (auto i = 0; i < NUM_VALUES; i++) {
+    h_a[i] = i;
+    h_b[i] = i * i;
+  }
+
+  auto d_a = cms::sycltools::make_device_unique<float[]>(NUM_VALUES, stream);
+  auto d_b = cms::sycltools::make_device_unique<float[]>(NUM_VALUES, stream);
+
+  stream.memcpy(d_a.get(), h_a.get(), NUM_VALUES * sizeof(float));
+  stream.memcpy(d_b.get(), h_b.get(), NUM_VALUES * sizeof(float));
+
+  unsigned int threadsPerBlock = std::min<unsigned int>(NUM_VALUES, workgroupSize);
+  unsigned int blocksPerGrid = (NUM_VALUES + threadsPerBlock - 1) / threadsPerBlock;
+
+  std::cerr << "block size: " << threadsPerBlock << " / " << stream.get_device().get_info<sycl::info::device::max_work_group_size>() << std::endl; 
+  std::cerr << "grid size:  " << blocksPerGrid   << std::endl;
+
+  auto d_c = cms::sycltools::make_device_unique<float[]>(NUM_VALUES, stream);
+  stream.submit([&](sycl::handler &cgh) {
+    auto d_a_get_ct0 = d_a.get();
+    auto d_b_get_ct1 = d_b.get();
+    auto d_c_get_ct2 = d_c.get();
+
+    cgh.parallel_for(sycl::nd_range<3>(sycl::range<3>(1, 1, blocksPerGrid) * sycl::range<3>(1, 1, threadsPerBlock),
+                                       sycl::range<3>(1, 1, threadsPerBlock)),
+                     [=](sycl::nd_item<3> item_ct1) {
+                       vectorAdd(d_a_get_ct0, d_b_get_ct1, d_c_get_ct2, NUM_VALUES, item_ct1);
+                     });
+  });
+
+  auto d_ma = cms::sycltools::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
+  auto d_mb = cms::sycltools::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
+  auto d_mc = cms::sycltools::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
+
+  sycl::range<3> threadsPerBlock3{workgroupRectA, workgroupRectB, 1};
+  sycl::range<3> blocksPerGrid3{(NUM_VALUES + threadsPerBlock3[0] - 1) / threadsPerBlock3[0], (NUM_VALUES + threadsPerBlock3[1] - 1) / threadsPerBlock3[1], 1};
+  stream.submit([&](sycl::handler &cgh) {
+    auto threadsPerGrid = blocksPerGrid3 * threadsPerBlock3;
+
+    auto d_a_get_ct0 = d_a.get();
+    auto d_b_get_ct1 = d_b.get();
+    auto d_ma_get_ct2 = d_ma.get();
+
+    cgh.parallel_for(
+        sycl::nd_range<3>(sycl::range<3>(threadsPerGrid.get(2), threadsPerGrid.get(1), threadsPerGrid.get(0)),
+                          sycl::range<3>(threadsPerBlock3.get(2), threadsPerBlock3.get(1), threadsPerBlock3.get(0))),
+        [=](sycl::nd_item<3> item_ct1) {
```

Comment:
Ok, now I'm just confused of how the `sycl::range`, `sycl::nd_range`, and `sycl::nd_item` work exactly.

Reply:
Welcome to the club :-(

As far as I understand:
  - `sycl::range` is a tuple, *e.g.* `64` (1-dimensional) or `4, 4, 4` (3-dimensional), and describes the "block" size (`blockDim`) OR the "grid" size (`gridDim`) OR the total size (`gridDim * blockDim`);
  - `sycl::nd_range` is two tuples, and describes the whole grid, like `<<<gridDim, blockDim>>>`, but it contains the *total* size and the *block* size; so `32 blocks x 32 threads` in CUDA are configred as `1024 global, 32 local`;
  - `sycl::nd_item` identifies one work item, like (`blockId, threadId`) in CUDA; it has accessors for the global, local and group ids, as well as global, local and group linearised indices.

Reply:
I think the conversion tools did a poor job here, I'll see if I can clean things up a bit.